src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs

```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
         /// <inheritdoc />
         public async Task<string> SetupAsync(SetupContext context)
         {
+            ShellSettings shellSettings = null;
             var initialState = context.ShellSettings.State;
             try
             {
-                var executionId = await SetupInternalAsync(context);
+                // Set the shell state to 'Initializing' to guarantee that any subsequent HTTP requests
+                // will result in a 'Service Unavailable' response while the tenant setup is in progress.
+                context.ShellSettings.AsInitializing();
 
-                if (context.Errors.Any())
+                (var executionId, shellSettings) = await SetupInternalAsync(context);
+
+                if (context.Errors.Count > 0)
                 {
                     context.ShellSettings.State = initialState;
-                    await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
+                    await _shellHost.ReloadShellContextAsync(shellSettings, eventSource: false);
+
+                    var shellScope = await _shellHost.GetScopeAsync(shellSettings);
+                    await shellScope.UsingAsync(async scope =>
+                    {
+                        var tenandSetupHandlers = scope.ServiceProvider.GetServices<ITenantSetupHandler>();
+
+                        await tenandSetupHandlers.InvokeAsync((handler) => handler.FailedAsync(context), _logger);
```

Comment:
Okay when at least you were calling these handlers below after the existing setup handlers that may add errors, this because the setup was almost done.

But here we also may just fail or retrieve an `Uninitialized` state without any features (of the setup recipe) enabled and then no tenant setup handlers to be resolved.

That's why for the `Failed` event I was thinking about a host level service, or at least a `DefaultTenantOnly` service belonging to a feature that would need to be part of the `Setup Features` , this to not let the tenant itself to do something if it failed to be setup.

But as said maybe okay if done after the existing setup handlers as you did previously.